### PR TITLE
feat(install): add tag matching for version tags

### PR DESCRIPTION
**Pull request by 0xYYY**
## Motivation

https://github.com/foundry-rs/foundry/issues/685
https://github.com/foundry-rs/foundry/issues/1555

### Tag Check
Since using `git ls-remote` to check tag will leave out commit hashes, we clone the repo and try to `git checkout` instead.

### Version Tags
If the supplied tag looks like a version tag, run `git tag` to get a lsit of tags and try to find matching ones and prompt the user to select the correct one.

## Test
Testing script
```bash
#!/usr/env/bin bash

# no tag
~/Repos/foundry/target/debug/forge install foundry-rs/forge-std

# branch
echo
~/Repos/foundry/target/debug/forge remove foundry-rs/forge-std >/dev/null
~/Repos/foundry/target/debug/forge install foundry-rs/forge-std@formatters

# commit
echo
~/Repos/foundry/target/debug/forge remove foundry-rs/forge-std >/dev/null
~/Repos/foundry/target/debug/forge install foundry-rs/forge-std@ea7c38a

# version tag (exact match)
echo
~/Repos/foundry/target/debug/forge remove foundry-rs/forge-std >/dev/null
~/Repos/foundry/target/debug/forge install foundry-rs/forge-std@v0.1.0

# version tag with only 1 match (but not exact match)
echo
~/Repos/foundry/target/debug/forge remove foundry-rs/forge-std >/dev/null
~/Repos/foundry/target/debug/forge install foundry-rs/forge-std@0.1

# version tag with multiple matches
echo
~/Repos/foundry/target/debug/forge install crytic/echidna@2.0
```

Output
```
Installing forge-std in "/home/aye/Repos/test/lib/forge-std" (url: Some("https://github.com/foundry-rs/forge-std"), tag: None)
    Installed forge-std

Installing forge-std in "/home/aye/Repos/test/lib/forge-std" (url: Some("https://github.com/foundry-rs/forge-std"), tag: Some("formatters"))
    Installed forge-std

Installing forge-std in "/home/aye/Repos/test/lib/forge-std" (url: Some("https://github.com/foundry-rs/forge-std"), tag: Some("ea7c38a"))
    Installed forge-std

Installing forge-std in "/home/aye/Repos/test/lib/forge-std" (url: Some("https://github.com/foundry-rs/forge-std"), tag: Some("v0.1.0"))
    Installed forge-std

Installing forge-std in "/home/aye/Repos/test/lib/forge-std" (url: Some("https://github.com/foundry-rs/forge-std"), tag: Some("0.1"))
Found a similar version tag: v0.1.0, do you want to use this insead? ([y]/n)
    Installed forge-std

Installing echidna in "/home/aye/Repos/test/lib/echidna" (url: Some("https://github.com/crytic/echidna"), tag: Some("2.0"))
There are multiple matching tags:
[0] SKIP AND USE ORIGINAL TAG
[1] v2.0.1
[2] v2.0.0-b2
[3] v2.0.0-b1
[4] v2.0.0
Please select a tag (0-4, default: 1): 4
[4] v2.0.0 selected
    Installed echidna
```

## Note
The current implementation follow the previous behavior where if a provided tag is incorrect, `forge install` will fail to install.

But we may want to consider falling back to master/main and give a clear warning to user, as proposed in https://github.com/foundry-rs/foundry/issues/1248. Since the user's intention is clear that they want to install this dependency and also deleting and installing the repo again may take some time if the repo is large.

So maybe providing warnings and instructions on what to do next (e.g. find correct tag, manually checkout and commit) is a better way. Would like to get some feedback on this. If agreed upon, I can make changes in this PR or do it in a separate one.

### CODE REVIEW

File: cli/src/cmd/forge/install.rs [hunk 1]
```diff
@@ -204,32 +218,130 @@ fn install_as_submodule(dep: &Dependency, libs: &Path, no_commit: bool) -> eyre:
...
+            eyre::bail!("{}", stderr.trim())
+        }
     }
 
-    Ok(())
+    Ok(tag)
+}
+
+/// disambiguate tag if it is a version tag
+fn match_tag(tag: &String, libs: &Path, target_dir: &str) -> eyre::Result<String> {
```
Comment (mattsse) [inline cli/src/cmd/forge/install.rs:266]: could you please add a brief note about the strategy here used to select tags?

File: cli/src/cmd/forge/install.rs [hunk 2]
```diff
@@ -204,32 +218,130 @@ fn install_as_submodule(dep: &Dependency, libs: &Path, no_commit: bool) -> eyre:
...
+        let matched_tag = candidates[0].clone();
+        print!(
+            "Found a similar version tag: {}, do you want to use this insead? ([y]/n)",
+            matched_tag
+        );
+        stdout().flush()?;
+        let mut input = String::new();
+        stdin().read_line(&mut input)?;
+        input = input.trim().to_lowercase();
+        if input.is_empty() || input == "y" || input == "yes" {
```
Comment (mattsse) [inline cli/src/cmd/forge/install.rs:312]: let's return here and get rid of the else branch this way

**Review by mattsse — COMMENTED** [review]
looking good,
but unsure if we want to start adding interactive support for selecting tags, no string feelings about this.

looks like this works, but this is rather hard to test against.

wdyt @onbjerg @gakonst

**Comment by mattsse** [comment]
yes, good catch! we should only ask if the terminal is tty and fail otherwise

File: cli/src/cmd/forge/install.rs (hunk 2 above)
Comment (0xYYY) [reply cli/src/cmd/forge/install.rs:312]: good suggestion, will do that insead

**Comment by 0xYYY** [comment]
Updated according to the reviews.

**Review by onbjerg — APPROVED** [review]
LGTM but will echo @mattsse's comment that a rustdoc for the `match_tag` function that briefly outlines the tag selection strategy would be nice (also needs a rebase)

**Comment by mattsse** [comment]
sg, needs another rebase though

**Review by gakonst — APPROVED** [review]
tested locally, lgtm

**Pull request merged**